tiled/adapters/xarray.py

```diff
@@ -1,223 +1,82 @@
-
-    def microstructure(self):
-        return None
-
-    def macrostructure(self):
-        if self._depth == 0:
-            coords = {k: v.structure() for k, v in self._coords.items()}
-        else:
-            coords = self._coords
-        return DataArrayMacroStructure(
-            variable=self._variable.structure(),
-            coords=coords,
-            coord_names=self._coord_names,
-            name=self._name,
-        )
-
-    def structure(self):
-        return DataArrayStructure(macro=self.macrostructure(), micro=None)
-
-    def read(self):
-        raise NotImplementedError
-        # return xarray.DataArray(...)
-
-
-class DatasetAdapter:
-    """
-    Wrap an xarray.Dataset
-    """
-
-    structure_family = "xarray_dataset"
-
-    def __init__(self, dataset):
-        self._dataset = dataset
-        self._data_vars = MapAdapter(
-            {
-                k: DataArrayAdapter.from_data_array(v)
-                for k, v in self._dataset.data_vars.items()
-            }
-        )
-        self._coords = MapAdapter(
-            {
-                k: DataArrayAdapter.from_data_array(v)
-                for k, v in self._dataset.coords.items()
-            }
-        )
-
-    @property
-    def metadata(self):
-        return {
-            "attrs": self._dataset.attrs,
-            "data_vars": {name: da.metadata for name, da in self._data_vars.items()},
-            "coords": {name: da.metadata for name, da in self._coords.items()},
-        }
-
-    def microstructure(self):
-        return None
-
-    def macrostructure(self):
+    def as_dataset(self):
+        # We do not stash the original dataset as state.
+        # We (re)construct one here, ensure that any filtering that was done
+        # is respected.
         data_vars = {}
-        for k, v in self._dataset.data_vars.items():
-            adapter = DataArrayAdapter.from_data_array(v, _depth=1)
-            data_vars[k] = {
-                "macro": adapter.macrostructure(),
-                "micro": adapter.microstructure(),
-            }
         coords = {}
-        for k, v in self._dataset.coords.items():
-            adapter = DataArrayAdapter.from_data_array(v, _depth=1)
-            coords[k] = {
-                "macro": adapter.macrostructure(),
-                "micro": adapter.microstructure(),
-            }
-        return DatasetMacroStructure(data_vars=data_vars, coords=coords)
+        for key, array_adapter in self.items():
+            if "xarray_data_var" in array_adapter.specs:
+                data_vars[key] = (
+                    array_adapter.macrostructure().dims,
+                    array_adapter.read(),
+                )
+            elif "xarray_coord" in array_adapter.specs:
+                coords[key] = (
+                    array_adapter.macrostructure().dims,
+                    array_adapter.read(),
+                )
+            else:
+                assert False, "Expected a spec"
```

Comment:
I think this should probably raise instead of assert

Reply:
It's true that `assert` should not be used for user input validation, but in this case the class is responsible for construction of the `array_adapters` and it should be impossible to reach this branch unless there is some bug internally to the class, so I think this is a correct use of `assert`.

Reply:
I think you can reach this path if you construct `DatasetAdapter` from an improper mapping just using `__init__`.

Reply:
You might actually want this if for example you are storing all the arrays separately instead of in a single netcdf and don't want to bother constructing the xarray object on the server side first which is probably how upload will have to work.

Reply:
Fair point. I have made it less strict.